trussworks/react-uswds
# feat: Add compositional Banner components

# Summary

Allow consumers to build custom Banners (say if a consumer needed a banner for a state government website)

This change is inspired by the DOL project which is not open source so I cannot link it here, but in short, we needed a custom Banner which GovBanner didn't satisfy the needs of.

I made the choice to leave the existing GovBanner in place because it's a nice convenience for consumers that want the default banner, and there's no easy way to introduce the breaking change workaround of removing it and then telling consumers to build a lengthy compositional component (see the [example](https://github.com/trussworks/react-uswds/blob/0e5368f9528f58fc0e431d1b8dc209b11ec722d5/src/components/banner/CustomBanner.stories.tsx#L18)) just to replace what they had. This way, there's no breaking changes, consumers are happy, and they have more options to boot.

## Related Issues or PRs

closes: #2183
closes: #1836 

## How To Test

1. Verify `GovBanner` has not changed
   - There is a snapshot jest test that does this for us. Note the only change is the use of the inline svg, which actually matches the USWDS banner component code _more closely_ than GovBanner previously did. 
   - Note that this means that the exported uswds lock svg isn't used. That is intentional, so that props can be passed in to the inline svg (title, description, so that consumers can pass in localized strings)
   - The lack of a Happo diff on GovBanner is also a good sign that this refactor breaks nothing
2. The new components are largely tested by GovBanner making use of them and GovBanner being tested, however each includes its own tests as well
3. Run storybook and view the `Components/Banner/Custom Banner` story

## Screenshots:

Custom banner storybook example (this one just changes some of the header text, but you can change _anything_! On the DOL project, we use the compositional structure to also allow us to internationalize _all_ of the text and aria content with i18n!:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/15805554/183538320-693d7b95-41d6-4339-9ffb-5859ed77e313.png">
